### PR TITLE
Create es index when a dataset published

### PR DESCRIPTION
**Pull request by frascuchon**
A basic integration with elasticsearch when a dataset is published. At this point, the elasticsearch index will be created based on dataset configuration.

A DB rollback is executed if some problem is found creating the index. If the error is raised from the commit statement, index creation should be reverted, or another mechanism to recreate the index should be provided through the API endpoint (and extra param with `force_recreate` or just `force`). But this won't be tackled in this PR.

Edit: Also, the elasticsearch client deps are reviewed. Now, the named 8x version is used. This allows us more control over the `es` client version to use, and it can suppress installation conflicts in an environment with a preview es client package installed.

### CODE REVIEW

File: src/argilla/server/elasticsearch.py [hunk 1]
```diff
@@ -53,4 +54,19 @@ def _field_mapping_for_annotation(self, annotation_task: Annotation):
             # See https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html
             return {"type": "text"}
         else:
-            raise ValueError(f"Annotation of type {settings_type} cannot be processed")
+            raise ValueError(f"ES mappings for Annotation of type {settings_type} cannot be generated")
+
+
+async def get_engine():
```
Comment (jfcalvo) [inline src/argilla/server/elasticsearch.py]: Looks like we are using search engine everywhere. Maybe we should use `get_search_engine` instead:

```suggestion
async def get_search_engine():
```

File: tests/conftest.py [hunk 2]
```diff
@@ -48,6 +51,29 @@ def db():
...
+    open_search = OpenSearch(hosts=settings.elasticsearch)
+
+    info = open_search.info(format="json")
+    version_info = info["version"]
+
+    return "distribution" not in version_info
+
+
+@pytest.fixture(scope="session")
+def es_config():
```
Comment (jfcalvo) [inline tests/conftest.py]:
```suggestion
def elasticsearch_config():
```

File: tests/server/api/v1/test_datasets.py [hunk 3]
```diff
@@ -1012,9 +1014,15 @@ def test_create_dataset_records_with_nonexistent_dataset_id(client: TestClient,
     assert db.query(Response).count() == 0
 
 
-def test_publish_dataset(client: TestClient, db: Session, admin_auth_header: dict):
+@pytest.mark.skipif(condition=not is_running_elasticsearch(), reason="Test only running with elasticsearch backend")
+def test_publish_dataset(
+    client: TestClient,
+    db: Session,
+    elasticsearch: Elasticsearch,
```
Comment (jfcalvo) [inline tests/server/api/v1/test_datasets.py:1021]: If we change this to `search_engine` on handlers maybe we should change it here too:

```suggestion
    search_engine: Elasticsearch,
```

Remember to change it for the rests of tests.

File: tests/server/api/v1/test_datasets.py [hunk 4]
```diff
@@ -1024,6 +1032,26 @@ def test_publish_dataset(client: TestClient, db: Session, admin_auth_header: dic
...
+
+@pytest.mark.skipif(condition=not is_running_elasticsearch(), reason="Test only running with elasticsearch backend")
+def test_publish_dataset_with_error_on_index_creation(
+    client: TestClient,
+    db: Session,
+    elasticsearch: Elasticsearch,
+    admin_auth_header: dict,
+):
+    dataset = DatasetFactory.create()
+    AnnotationFactory.create(dataset=dataset)
```
Comment (jfcalvo) [inline tests/server/api/v1/test_datasets.py]:
```suggestion
    AnnotationFactory.create(settings={"type": "invalid"}, dataset=dataset)
```

**Review by jfcalvo — APPROVED** [review]

File: tests/server/api/v1/test_datasets.py (hunk 3 above)
Comment (frascuchon) [reply tests/server/api/v1/test_datasets.py:1021]: This is not the search engine, is just the official elasticsearch client used to check test results.

**Review by jfcalvo — APPROVED** [review]

**Pull request merged**